### PR TITLE
CONAN_COLOR_DISPLAY forces ANSI sequences on CI

### LINKED ISSUE #4718: Conan doesn't print coloured output in Gitlab CI
To help us debug your issue please explain:

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

Using Conan v1.11.2 on `lasote/conangcc7` and Windows 10, we don't get coloured output in our Gitlab CI build jobs. We have even tried setting the `CONAN_COLOR_DISPLAY` variable to 1, and it still doesn't work.

### PR DESCRIPTION
**Pull request by ytimenkov**
CONAN_COLOR_DISPLAY if set to 1 forces colorama to print ANSI sequences
even when output is not a tty.

Fixes #4718

Changelog: Fix: CONAN_COLOR_DISPLAY forces ANSI sequences on CI

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

### CODE REVIEW

**Comment by ytimenkov** [comment]
Hmm... revision_list_test.py fails on develop as well...

File: conans/client/conan_api.py [hunk 1]
```diff
@@ -173,19 +173,22 @@ def instance_remote_manager(requester, cache, user_io, hook_manager):
...
+                break
+        else:
+            force_color = False
+            color = hasattr(sys.stdout, "isatty") and sys.stdout.isatty()
+
+        if color:
             import colorama
-            if get_env("PYCHARM_HOSTED"):  # in PyCharm disable convert/strip
+            if force_color:
                 colorama.init(convert=False, strip=False)
```
Comment (lasote) [inline conans/client/conan_api.py]: Previously this was only done when `PYCHARM_HOSTED` but now also with `CONAN_COLOR_DISPLAY`, right? I would need some explanation about this change and how it fixes the issue.

**Comment by ytimenkov** [comment]
> I'm not sure I understand this, also, about your comment in the issue:
> 
> Unfortunately this kicked my butt: conan --version and conan remote list --raw became not parser-friendly disappointed
> 
> So, is this PR finally working or not?
Let me see...

Essentially you get what you ask for: if you ask to force colors, you get colors.

Including in the output of `conan --version` command. I worked this around by running commands which parse output with explicit `CONAN_COLOR_DISPLAY=0`.
Most commands are not affected because they have handy json output which writes to file and 100% machine-readable.

The only question is about coloring of `conan remote list --raw` output. I would expect parameters like `--raw` or `--porcelain` to output machine-readable data which should not contain any ANSI sequences under any circumstances.

File: conans/client/conan_api.py (hunk 1 above)
Comment (ytimenkov) [reply conans/client/conan_api.py]: That is the fix: if user sets `CONAN_COLOR_DISPLAY` to `1` it should behave exactly as when `PYCHARM_HOSTED` present. (Research in my comment to the issue https://github.com/conan-io/conan/issues/4718#issuecomment-478533897)

Major complaint is that setting `CONAN_COLOR_DISPLAY=1` doesn't really force colors (as I as a user would expect) because colorama does check for tty internally and essentially this variable is being ignored.

Now when either of those variables are set to 1 conan forces ANSI sequences in the output regardless whether it's a tty or not. When it's set to 0 the effect is the opposite: no colors even when stdout is a tty.
That's why 2 variables: `force_color` and `color`. If `force_color` is not set then colorama is initialized as before with default parameters.

**Comment by lasote** [comment]
Yes, I agree with that. But as it is totally broken in the Conan side, I'm afraid we cannot force the ANSI sequences when `CONAN_COLOR_DISPLAY=1` even without tty or we could be breaking something.

**Comment by ytimenkov** [comment]
It's not totally broken. It just that conan prints ansi sequences but Colorama replaced stdout which strips them.

**Comment by lasote** [comment]
I meant "broken", the fact that Conan, for the `--raw` and pristine, etc, is not excluding the coloring.

**Comment by ytimenkov** [comment]
So what do you suggest? I don't see it as a severely broken. There are not that many commands which should output data without ANSI sequences and it's just a matter of changing `user_io.out.info` to `user_io.out.write(ln)` as it is done for build_order).

**Comment by jgsogo** [comment]
I've added a test to cover all the use cases, there were so many `if/else` and I wasn't able to follow them. The test passes with the sources in `develop` and with these changes it is failing (it can be good or not).

The behavior that has changed:

Given:
 * assert sys.stdout.isatty() is False
 * assert 'CONAN_COLOR_DISPLAY' not in os.environ

The call to `colorama_initialize()`:
 * now returns `True`
 * now `colorama.init` is called

TL;DR: for non-terminals now we are getting colors too when the environment variable is not set (it is like defaulting to colors).

**Comment by ytimenkov** [comment]
What about `PYCHARM_HOSTED`? Do you run it from pycharm?
With my changes they became synonyms...

**Comment by jgsogo** [comment]
I didn't put anything explicit because it was not affecting the behavior. Regarding the `PYCHARM_HOSTED` everything works the same.

That variable only affects to the parameters passed to `colorama.init`:
 * if it is "True", then `colorama.init(convert=False, strip=False)`
 * if it is "False", then `colorama.init()` (which is `colorama.init(convert=None, strip=None)`)

**Comment by ytimenkov** [comment]
I've debugged the failing `test_tty_false` and it sets `PYCHARM_HOSTED_VALUE=1` in both  expansions.

It also seems that `_` and `pycharm_hosted` parameters are reversed, because debugger shows type of `pycharm_hosted` being `<MagicMock name='isatty' id='90357744'>` while `_` is just `True` or `False`

**Comment by ytimenkov** [comment]
Also `with environment_append({PYCHARM_HOSTED: PYCHARM_HOSTED_VALUE}):` won't allow test to pass because it also forces color if `CONAN_COLOR_DISPLAY` is unset (the first check).

**Comment by ytimenkov** [comment]
Digging into history and trying to follow control paths I think we don't need to look at `PYCHARM_HOSTED` variable at all.
You're right, in original commit 75d618ebea28687ecb5f779568b995899e5bf5e1 by @lachlankrautz, `PYCHARM_HOSTED` was taken into account only when `CONAN_COLOR_DISPLAY` was set to 1.
In this PR `convert=False, strip=False` are passed to colorama when `CONAN_COLOR_DISPLAY` is set to `1`.

**Comment by jgsogo** [comment]
🤦‍♂ You are totally right, @ytimenkov, the test was wrong. I've fixed it and strikethrough my previous comments.

I keep trying to resume the implementation you are proposing. With this fix, the behavior in your implementation changes a little bit:
 * `tty` makes no difference. The behavior is the same if it is returning `True` or `False`
 * whenever we are calling `colorama.init` it is always passed `convert=False` and `strip='not-set'`
 * We are calling `colorama.init` when:
   * `CONAN_COLOR_DISPLAY` equals `1`
   * `CONAN_COLOR_DISPLAY` is not set and `pycharm_hosted` is `True`
 * We are NOT calling `colorama.init` when:
   * `CONAN_COLOR_DISPLAY` equals `0`
   * `CONAN_COLOR_DISPLAY` is not set and `pycharm_hosted` is `False`

**Comment by jgsogo** [comment]
Regarding the original issue, if `CONAN_COLOR_DISPLAY` is set to `1/True` the released behavior is the same as the one proposed here, so if it was not printing colors, it won't be.

There is something to mention in this issue that could be root of the problem. As soon as Conan is imported, we are importing `tqdm` which in turn is calling `colorama.init()` for  Windows (https://github.com/tqdm/tqdm/blob/d69eb9cf2c9cbfe8816251e349e2f00666314132/tqdm/_utils.py#L34) and it is interfering with our implementation. And maybe some other packages are using this same approach...

Maybe we need to fix the `tqdm` issue in their package (they may need a `tqdm.init` explicit function) before targeting our problem. We cannot risk Conan stability (now it is working in _every_ terminal) to get colors in some CI systems. We have to be very careful with this.

**Comment by ytimenkov** [comment]
One thought: we can add another value to `CONAN_COLOR_DISPLAY`, for example `ANSI` or `FORCE` as a 3rd option to force ANSI sequences even on non-ttys (and drop `PYCHARM_HOSTED` variable).

The only concern is that if old version runs where `CONAN_COLOR_DISPLAY` is set to non-integer value user gets `ValueError` immediately. But running old version of conan anyways leads to many problems (e.g. settings migrations).

**Comment by saimusdev** [comment]
> Hi all. We are facing here very important issues, even with external deps like tqdm. In my opinion too much only for showing colored output in very concrete circumstances, I'm sorry but unless you say the opposite, I vote for closing this.
> I really really appreciate your time and help @ytimenkov, honestly, but we have to be pragmatic and keep the maintenance of the project at a sane level.

Could you provide a current estimation on the cost to implement this feature / change request. I do think it's very useful to provide colorized output in a CI environment as it can greatly help debug conan builds, just as they do on a local (tty) execution.

**Comment by ytimenkov** [comment]
I actually out of sudden though what we can introduce a new variable instead `COnan_force_ansi_colors` or so which does exactly as what it name says (if set to true-ish, otherwise ignored).
No change to any existing functionality, no break.
(I think in this case we don't even need to initialize colorama).

**Comment by saimusdev** [comment]
> I actually out of sudden though what we can introduce a new variable instead `COnan_force_ansi_colors` or so which does exactly as what it name says (if set to true-ish, otherwise ignored).
> No change to any existing functionality, no break.
> (I think in this case we don't even need to initialize colorama).

@ytimenkov Is this something you can easily implement?
@lasote What do you think about this solution, could this be achievable and something to consider to include to Conan?

**Comment by ytimenkov** [comment]
To implement is not a problem, maybe five lines of code. To verify is trickier: at least windows and posix machines needed. Most work is writing proper tests (although I think the feature itself is fairly straightforward).

I'm on parental till August and unfortunately have extremely little free time.
